asm/amd/regs_state.go

```diff
@@ -3,64 +3,310 @@
+		if reg != x86asm.RIP {
+			fmt.Printf("                               -> | %6s = %s\n", reg, v.String())
+		}
+	}
+	r.regs[idx] = v
+}
+
+func (r *RegsState) Get(reg x86asm.Reg) variable.U64 {
+	idx, sz := regIndexWithSize(reg)
+	res := r.regs[idx]
+	if sz != size64 {
+		res = variable.Crop(res, sz)
+	}
+	return res
 }
 
-func (r *RegsState) Get(reg x86asm.Reg) (value, loadedFrom uint64) {
-	return r.regs[regIndex(reg)].Value, r.regs[regIndex(reg)].LoadedFrom
+type Interpreter struct {
+	Regs        RegsState
+	code        []byte
+	CodeAddress variable.U64
+	pc          int
 }
 
-type regState struct {
-	LoadedFrom uint64
-	Value      uint64
+func NewInterpreter(code []byte) Interpreter {
+	it := Interpreter{code: code, CodeAddress: variable.Var("code address")}
+	it.initRegs()
+	return it
+}
+
+func (i *Interpreter) initRegs() {
+	i.Regs.regs[0] = variable.Var("invali reg")
+	i.Regs.regs[regIndex(x86asm.RAX)] = variable.Var("initial RAX")
+	i.Regs.regs[regIndex(x86asm.RCX)] = variable.Var("initial RCX")
+	i.Regs.regs[regIndex(x86asm.RDX)] = variable.Var("initial RDX")
+	i.Regs.regs[regIndex(x86asm.RBX)] = variable.Var("initial RBX")
+	i.Regs.regs[regIndex(x86asm.RSP)] = variable.Var("initial RSP")
+	i.Regs.regs[regIndex(x86asm.RBP)] = variable.Var("initial RBP")
+	i.Regs.regs[regIndex(x86asm.RSI)] = variable.Var("initial RSI")
+	i.Regs.regs[regIndex(x86asm.RDI)] = variable.Var("initial RDI")
+	i.Regs.regs[regIndex(x86asm.R8)] = variable.Var("initial R8")
+	i.Regs.regs[regIndex(x86asm.R9)] = variable.Var("initial R9")
+	i.Regs.regs[regIndex(x86asm.R10)] = variable.Var("initial R10")
+	i.Regs.regs[regIndex(x86asm.R11)] = variable.Var("initial R11")
+	i.Regs.regs[regIndex(x86asm.R12)] = variable.Var("initial R12")
+	i.Regs.regs[regIndex(x86asm.R13)] = variable.Var("initial R13")
+	i.Regs.regs[regIndex(x86asm.R14)] = variable.Var("initial R14")
+	i.Regs.regs[regIndex(x86asm.R15)] = variable.Var("initial R15")
+	i.Regs.regs[regIndex(x86asm.RIP)] = variable.Var("initial RIP")
+}
+
+func (i *Interpreter) Loop() error {
+	return i.LoopWithBreak(nil)
+}
+func (i *Interpreter) LoopWithBreak(breakLoop func(op x86asm.Inst) bool) error {
+	for j := 0; j < 137; j++ {
+		op, err := i.Step()
+		if err != nil {
+			return err
+		}
+		if breakLoop != nil && breakLoop(op) {
+			return nil
+		}
+	}
+	return errors.New("interpreter loop bound")
+}
+func (i *Interpreter) Step() (x86asm.Inst, error) {
+	rem := i.code[i.pc:]
+	if len(rem) == 0 {
+		return x86asm.Inst{}, io.EOF
+	}
+	if ok, insnLen := DecodeSkippable(rem); ok {
+		i.pc += insnLen
+		return x86asm.Inst{Op: x86asm.NOP}, nil
+	}
+	inst, err := x86asm.Decode(rem, 64)
+	if err != nil {
+		return x86asm.Inst{}, fmt.Errorf("failed to decode instruction at 0x%x : %w",
+			i.pc, err)
+	}
+	if debugPrinting {
+		fmt.Printf(" | %4x %s\n", i.pc, inst.String())
+	}
+	i.pc += inst.Len
+	i.Regs.Set(x86asm.RIP, variable.Add(i.CodeAddress, variable.Imm(uint64(i.pc))))
+
+	if inst.Op == x86asm.RET {
```

Comment:
I would do `switch inst.Op` here as that is what the individual `if` blocks test on.